imatlab_export_fig/imatlab_set_exporter.m

```diff
@@ -0,0 +1,81 @@
+function imatlab_set_exporter(exporter)
+%
+% imatlab_set_exporter(exporter)
+%   sets exporter for imatlab, useful for inline graphics in notebook mode
+% Ex:
+% imatlab_set_exporter(); % defaults to fig2plotly
+% imatlab_set_exporter('fig2plotly');
+% imatlab_set_exporter('print'); % prints to png
+% imatlab_set_exporter('static'); prints to png
+
+
+  if nargin < 1
+    exporter = 'fig2plotly';
+  end
+
+  if ~any( strcmp(exporter, {'fig2plotly', 'print', 'static'}))
+    error('imatlab:imatlab_set_exporter:bad_exporter', 'Invalid exporter specified');
+  end
+
+
+  switch exporter
+    case 'fig2plotly'
+
+      setenv('IMATLAB_EXPORT_FIG', func2str(@(h) eval([ ...
+             'fig2plotly(h, ''filename'', [tempname(''.''), ''.html''], ', ...
+             '''offline'', true, ''open'', false); ', ...
+             'close(h);'])));
+
+  case {'print', 'static'}
+
+    setenv('IMATLAB_EXPORT_FIG', func2str(@(h) eval([ ...
+    'print(h, [tempname(''.''), ''.png''], ''-dpng'');', ...
+    'close(h);'])));
+
+end
+
+
+% function imatlab_set_exporter(exporter, fileType, varargin)
+%  fileTypes = { '-djpeg', '-dpng', '-dtiff', '-dtiffn', '-dmeta', '-dbmpmono', ...
+%                '-dbmp', '-dbmp16m', '-dbmp256', '-dhdf', '-dpbm', '-dpbmraw', ...
+%                '-dpcxmono', '-dpcx24b', '-dpcx256', '-dpcx16', '-dpgm', '-dpgmraw', ...
+%                '-dppm', '-dppmraw', '-dpdf', '-deps', '-depsc', '-deps2', '-depsc2', ...
+%                '-dmeta', '-dsvg', '-dps', '-dpsc', '-dps2', '-dpsc2'; ...
+%                '.jpg', 'png', 'tif', 'tif', 'emf', '.bmp', '.bmp', '.bmp', '.bmp', ...
+%                '.hdf', '.pbm', '.pbm', '.pcx', '.pcx.', '.pcx.', '.pcx', ...
+%                '.pgm', '.pgm', '.ppm', '.ppm', '.pdf', '.eps', '.eps', '.eps', '.eps', ...
+%                '.emf', '.svg', '.ps', '.ps', '.ps', '.ps'};
+
+
+
+%    if exist('fileType', 'var') && ~isempty(fileType)
+%        fileExt = fileTypes{2, find(strcmp(fileType, fileTypes))};
+%    else
+%        fileType = '-dpng';
+%        fileExt = '.png';
+%    end
+%
+%
+%    strOptions = [];
+%    if ~isempty(varargin)
+%        for i = 1 : numel(varargin)
+%            strOptions = [strOptions, ', ', varargin{i}];
+%        end
+%    end
+%
+%    printStr = sprintf( 'print(h, [tempname(''.''), ''%s''], ''%s'' %s );', fileExt, fileType, strOptions);
```

Comment:
Given the issues with JuyterHub dropping environment variables by default, this may be the better route.  The spec would be something like:

After a cell is executed, the kernel checks if a function named `imatlab_exporter` exists in the path.  If it does, the kernel creates a new temporary directory, cd's into it (1), and calls with no arguments.  The function must return a cell array of chars (2), which are individual filenames to be found in the directory.  These filenames must have extension `.html` or `.png` (others will be dropped); they will be loaded into the notebook.

Rationale for (1): `fig2plotly` always creates its output in the current directory, even if it is passed an absolute path (which is interpreted as a subfolder of the current directory).  Letting the kernel handle this also makes cleanup easier.

Rationale for (2): Doing the looping over handles at the MATLAB level makes it easier to decide and propagate back to the kernel the order in which the figures should be loaded into the notebook.

imatlab includes a default implementation of `imatlab_exporter` (next to the .py files -- the folder is added to the path by default by the kernel).  This function, by default, does nothing.  After calling `imatlab_exporter('plotly')`, it becomes a plotly exporter.  After calling `imatlab_exporter('static', ...)`, it becomes a static exporter.  After calling `imatlab_exporter('')`, it returns to doing nothing.  (The function stores the required internal state using the `persistent` mechanism.)

Now someone just needs to implement this :-)